File: 127-word-ladder/word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        queue<pair<string,int>> q;
        q.push({beginWord, 1});
        unordered_set<string> st(wordList.begin(), wordList.end());

        st.erase(beginWord);


        while(!q.empty())
        {


            string currWord = q.front().first;
            int step = q.front().second;
            q.pop();

            if(currWord == endWord) return step;

            for(int i = 0;i< currWord.length();i++)
            {
                char og = currWord[i];

                for(char ch='a' ; ch <= 'z';ch++)
                {
                    currWord[i] = ch;
                    if(st.find(currWord) != st.end()) {
                        st.erase(currWord);
                        q.push({currWord, step +1});
                        
                    }
                    
                }
                currWord[i] = og;
            }
            
        }
        return 0;
    }
};

File: 127-word-ladder/word-ladder.java
class Solution {
    
    class Pair{
        String first;
        int second;
        Pair(String first, int second)
        {
            this.first = first;
            this.second = second;
        }
    }
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set = new HashSet<>();
        for(String s : wordList)
        {
            set.add(s);
        }
        if(set.contains(beginWord))
        {
            set.remove(beginWord);
        }
        if(!set.contains(endWord))
        {
            return 0;
        }

        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(beginWord, 1));
 

        while(!q.isEmpty())
        {
            Pair curr = q.poll();
            String currentWord = curr.first;
            if(currentWord.equals(endWord)) return curr.second;
            for(int i = 0 ;i <currentWord.length() ; i++)
            {
                char[] newWord = currentWord.toCharArray();
                for(int j= 0; j<26 ; j++)
                {

                    newWord[i] = (char)(j + 97);
                    String temp = new String(newWord); 
                    if(set.contains(temp))
                    {
                        q.add(new Pair(temp , curr.second + 1));
                        set.remove(temp);
                    }
                }

            }
      

        }
          return 0;
    }
}